File: tests/core/test_lda_projection.py
# SPDX-License-Identifier: MIT OR Apache-2.0
# Copyright (c) 2025 John William Creighton (s243a)
#
# Unit tests for LDA projection Python module

import sys
import os
import tempfile
import numpy as np

# Add the source directory to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '../../src/unifyweaver/targets/python_runtime'))

from projection import (
    LDAProjection,
    compute_weighted_centroid,
    compute_W,
    save_W,
    load_W
)


def test_weighted_centroid():
    """Test iterative weighted centroid computation."""
    print("Test: Weighted centroid computation")

    # Create simple test case: 3 vectors, one outlier
    questions = np.array([
        [1.0, 0.0, 0.0],      # Representative
        [0.9, 0.1, 0.0],      # Similar to first
        [0.0, 1.0, 0.0],      # Outlier
    ])

    centroid, weights = compute_weighted_centroid(questions, max_iter=3)

    # Centroid should be closer to first two vectors
    assert np.dot(centroid, questions[0]) > np.dot(centroid, questions[2]), \
        "Centroid should be closer to representative vectors"

    # Weights should sum to 1
    assert abs(np.sum(weights) - 1.0) < 1e-6, "Weights should sum to 1"

    # First two vectors should have higher weight than outlier
    assert weights[0] > weights[2], "Representative vector should have higher weight"
    assert weights[1] > weights[2], "Similar vector should have higher weight"

    print("  ✓ Weighted centroid computed correctly")


def test_compute_W_simple():
    """Test W matrix computation with simple clusters."""
    print("Test: W matrix computation (simple)")

    # Create simple test case
    d = 4  # 4-dimensional embeddings
    np.random.seed(42)

    # Create 2 clusters
    clusters = [
        # Cluster 1: answer is [1, 0, 0, 0], questions around [0, 1, 0, 0]
        (
            np.array([1.0, 0.0, 0.0, 0.0]),  # answer
            np.array([
                [0.0, 0.9, 0.1, 0.0],       # questions
                [0.0, 0.8, 0.2, 0.0],
                [0.1, 0.85, 0.05, 0.0],
            ])
        ),
        # Cluster 2: answer is [0, 0, 1, 0], questions around [0, 0, 0, 1]
        (
            np.array([0.0, 0.0, 1.0, 0.0]),  # answer
            np.array([
                [0.0, 0.0, 0.1, 0.9],       # questions
                [0.0, 0.1, 0.0, 0.9],
                [0.0, 0.0, 0.2, 0.8],
            ])
        ),
    ]

    W = compute_W(clusters, lambda_reg=1.0, ridge=1e-6)

    assert W.shape == (d, d), f"W should be {d}x{d}"

    # Project a query from cluster 1's question space
    q1 = np.array([0.0, 0.9, 0.1, 0.0])
    q1_proj = W @ q1

    # Project a query from cluster 2's question space
    q2 = np.array([0.0, 0.0, 0.1, 0.9])
    q2_proj = W @ q2

    # q1_proj should be more similar to cluster 1's answer
    sim1_a1 = np.dot(q1_proj, clusters[0][0]) / (np.linalg.norm(q1_proj) * np.linalg.norm(clusters[0][0]))
    sim1_a2 = np.dot(q1_proj, clusters[1][0]) / (np.linalg.norm(q1_proj) * np.linalg.norm(clusters[1][0]))

    print(f"    q1 projected similarity to a1: {sim1_a1:.4f}")
    print(f"    q1 projected similarity to a2: {sim1_a2:.4f}")

    # Note: This might not always pass with simple test data
    # The test mainly verifies that compute_W runs without errors

    print("  ✓ W matrix computed successfully")


def test_projection_class():
    """Test LDAProjection class."""
    print("Test: LDAProjection class")

    # Create a simple test W matrix (rotation + scaling)
    d = 4
    theta = np.pi / 4  # 45 degree rotation in first two dimensions
    W = np.eye(d)
    W[0, 0] = np.cos(theta)
    W[0, 1] = -np.sin(theta)
    W[1, 0] = np.sin(theta)
    W[1, 1] = np.cos(theta)

    # Save to temp file
    with tempfile.NamedTemporaryFile(suffix='.npy', delete=False) as f:
        temp_path = f.name
        np.save(temp_path, W)

    try:
        # Load via class
        projection = LDAProjection(temp_path, embedding_dim=d)

        # Test single projection
        query = np.array([1.0, 0.0, 0.0, 0.0])
        projected = projection.project(query)

        expected = np.array([np.cos(theta), np.sin(theta), 0.0, 0.0])
        assert np.allclose(projected, expected, atol=1e-6), \
            f"Projection mismatch: got {projected}, expected {expected}"

        print("  ✓ Single projection works")

        # Test batch projection
        queries = np.array([
            [1.0, 0.0, 0.0, 0.0],
            [0.0, 1.0, 0.0, 0.0],
        ])
        projected_batch = projection.project_batch(queries)

        assert projected_batch.shape == (2, d), f"Batch shape mismatch"
        assert np.allclose(projected_batch[0], expected, atol=1e-6), "Batch projection[0] mismatch"

        print("  ✓ Batch projection works")

        # Test similarity
        doc = np.array([np.cos(theta), np.sin(theta), 0.0, 0.0])
        sim = projection.projected_similarity(query, doc)
        assert abs(sim - 1.0) < 1e-6, f"Expected similarity 1.0, got {sim}"

        print("  ✓ Projected similarity works")

    finally:
        os.unlink(temp_path)


def test_save_load_W():
    """Test saving and loading W matrix."""
    print("Test: Save/Load W matrix")

    W = np.random.randn(4, 4)

    # Test .npy format
    with tempfile.NamedTemporaryFile(suffix='.npy', delete=False) as f:
        npy_path = f.name

    try:
        save_W(W, npy_path)
        W_loaded = load_W(npy_path)
        assert np.allclose(W, W_loaded), "NPY save/load mismatch"
        print("  ✓ NPY format works")
    finally:
        os.unlink(npy_path)

    # Test .json format
    with tempfile.NamedTemporaryFile(suffix='.json', delete=False) as f:
        json_path = f.name

    try:
        save_W(W, json_path)
        W_loaded = load_W(json_path)
        assert np.allclose(W, W_loaded, atol=1e-10), "JSON save/load mismatch"
        print("  ✓ JSON format works")
    finally:
        os.unlink(json_path)


def test_batch_similarity():
    """Test batch similarity computation."""
    print("Test: Batch similarity")

    d = 4
    W = np.eye(d)  # Identity matrix for simple test

    with tempfile.NamedTemporaryFile(suffix='.npy', delete=False) as f:
        temp_path = f.name
        np.save(temp_path, W)

    try:
        projection = LDAProjection(temp_path, embedding_dim=d)

        queries = np.array([
            [1.0, 0.0, 0.0, 0.0],
            [0.0, 1.0, 0.0, 0.0],
        ])
        docs = np.array([
            [1.0, 0.0, 0.0, 0.0],
            [0.0, 1.0, 0.0, 0.0],
            [0.0, 0.0, 1.0, 0.0],
        ])

        sims = projection.projected_similarity_batch(queries, docs)

        assert sims.shape == (2, 3), f"Similarity matrix shape mismatch: {sims.shape}"

        # With identity W, q1 should be most similar to d1, q2 to d2
        assert abs(sims[0, 0] - 1.0) < 1e-6, "q1-d1 should be 1.0"
        assert abs(sims[1, 1] - 1.0) < 1e-6, "q2-d2 should be 1.0"
        assert abs(sims[0, 1]) < 1e-6, "q1-d2 should be 0.0"

        print("  ✓ Batch similarity works")

    finally:
        os.unlink(temp_path)


def create_test_w_matrix(output_path, dim=384):
    """Create a test W matrix for integration testing.

    Creates a simple matrix that rotates the embedding space slightly
    to simulate learned projection.
    """
    print(f"Creating test W matrix ({dim}x{dim}) at {output_path}")

    # Create a simple transformation:
    # - Small rotation in first few dimensions
    # - Identity in remaining dimensions
    W = np.eye(dim)

    # Add small rotation in dimensions 0-1
    theta = 0.1  # Small angle
    W[0, 0] = np.cos(theta)
    W[0, 1] = -np.sin(theta)
    W[1, 0] = np.sin(theta)
    W[1, 1] = np.cos(theta)

    save_W(W, output_path)
    print(f"  ✓ Test W matrix saved to {output_path}")

    return W


def run_all_tests():
    """Run all tests."""
    print("\n=== LDA Projection Python Tests ===\n")

    test_weighted_centroid()
    test_compute_W_simple()
    test_projection_class()
    test_save_load_W()
    test_batch_similarity()

    print("\nAll tests passed!")


if __name__ == '__main__':
    import argparse

    parser = argparse.ArgumentParser(description='LDA Projection Tests')
    parser.add_argument('--create-test-matrix', type=str, metavar='PATH',
                       help='Create a test W matrix at the specified path')
    parser.add_argument('--dim', type=int, default=384,
                       help='Dimension for test matrix (default: 384)')

    args = parser.parse_args()

    if args.create_test_matrix:
        create_test_w_matrix(args.create_test_matrix, args.dim)
    else:
        run_all_tests()

File: tests/core/test_lda_projection.pl
% SPDX-License-Identifier: MIT OR Apache-2.0
% Copyright (c) 2025 John William Creighton (s243a)
%
% Integration tests for LDA projection through component registry

:- encoding(utf8).

% Import only what we need from component_registry to avoid conflicts
:- use_module('../../src/unifyweaver/core/component_registry', [
    component_type/4,
    list_types/2,
    list_components/2,
    declare_component/4,
    retract_component/2,
    component/4
]).

% Load lda_projection module (it auto-registers itself)
:- use_module('../../src/unifyweaver/runtime/lda_projection', [
    type_info/1,
    validate_config/1
]).

%% ============================================
%% Test Helpers
%% ============================================

run_tests :-
    format('~n=== LDA Projection Integration Tests ===~n~n'),
    (   setup_test_matrix(TempFile)
    ->  (   run_all_tests(TempFile)
        ->  cleanup_test(TempFile),
            format('~nAll tests passed!~n')
        ;   cleanup_test(TempFile),
            format('~nTests FAILED~n'),
            halt(1)
        )
    ;   format('Failed to create test matrix~n'),
        halt(1)
    ).

run_all_tests(TempFile) :-
    test_type_info,
    test_type_registration,
    test_config_validation,
    test_component_declaration(TempFile),
    test_projection_invocation(TempFile).

assert_true(Goal, TestName) :-
    (   call(Goal)
    ->  format('  ~c ~w~n', [10003, TestName])  % checkmark
    ;   format('  X ~w FAILED~n', [TestName]),
        fail
    ).

assert_false(Goal, TestName) :-
    (   \+ call(Goal)
    ->  format('  ~c ~w~n', [10003, TestName])  % checkmark
    ;   format('  X ~w FAILED (expected false)~n', [TestName]),
        fail
    ).

%% ============================================
%% Setup: Create Test W Matrix
%% ============================================

setup_test_matrix(TempFile) :-
    format('Setting up test W matrix...~n'),
    % Create a simple 4x4 identity W matrix for testing
    get_time(Now),
    NowInt is round(Now * 1000000),
    format(atom(TempFile), '/tmp/test_W_~d.npy', [NowInt]),
    format(atom(PythonCode),
'import numpy as np
W = np.eye(4)  # Simple identity matrix
np.save("~w", W)
print("OK")
', [TempFile]),
    process_create(path(python3), ['-c', PythonCode],
                  [stdout(pipe(Out)), stderr(null), process(Proc)]),
    read_string(Out, _, Result),
    close(Out),
    process_wait(Proc, Status),
    (   Status = exit(0), sub_string(Result, _, _, _, "OK")
    ->  format('  Test matrix created: ~w~n~n', [TempFile])
    ;   format('  Failed to create test matrix~n'),
        fail
    ).

cleanup_test(TempFile) :-
    (   exists_file(TempFile)
    ->  delete_file(TempFile)
    ;   true
    ).

%% ============================================
%% Test: Type Info
%% ============================================

test_type_info :-
    format('Test: LDA projection type info~n'),
    lda_projection:type_info(Info),
    Info = info(name(Name), version(Version), _, _, _, _, _),
    assert_true(atom(Name), 'type has name'),
    assert_true(atom(Version), 'type has version'),
    format('~n').

%% ============================================
%% Test: Type Registration
%% ============================================

test_type_registration :-
    format('Test: LDA projection type registration~n'),
    % lda_projection registers itself on load
    assert_true(
        component_type(runtime, lda_projection, lda_projection, _),
        'lda_projection type registered in runtime category'
    ),
    list_types(runtime, Types),
    assert_true(member(lda_projection, Types), 'lda_projection in runtime types'),
    format('~n').

%% ============================================
%% Test: Configuration Validation
%% ============================================

test_config_validation :-
    format('Test: LDA projection config validation~n'),

    % Valid config should pass
    assert_true(
        lda_projection:validate_config([model_file('/tmp/test.npy')]),
        'valid config passes'
    ),

    % Config with optional params should pass
    assert_true(
        lda_projection:validate_config([
            model_file('/tmp/test.npy'),
            embedding_dim(384),
            lambda_reg(1.0),
            ridge(0.000001)
        ]),
        'config with optional params passes'
    ),

    % Missing model_file should fail
    assert_false(
        catch(lda_projection:validate_config([]), _, fail),
        'missing model_file fails'
    ),

    format('~n').

%% ============================================
%% Test: Component Declaration
%% ============================================

test_component_declaration(TempFile) :-
    format('Test: LDA projection component declaration~n'),

    % Declare a test component
    declare_component(runtime, test_projection, lda_projection, [
        model_file(TempFile),
        embedding_dim(4),
        initialization(lazy)
    ]),

    assert_true(
        component(runtime, test_projection, lda_projection, _),
        'component declared successfully'
    ),

    list_components(runtime, Names),
    assert_true(member(test_projection, Names), 'test_projection in components list'),

    format('~n').

%% ============================================
%% Test: Projection Invocation
%% ============================================

test_projection_invocation(TempFile) :-
    format('Test: LDA projection invocation~n'),

    % Test single query projection (through direct module call)
    QueryEmb = [1.0, 0.0, 0.0, 0.0],

    % Using the module's invoke_component directly
    Config = [model_file(TempFile), embedding_dim(4)],

    (   catch(
            lda_projection:invoke_component(test_projection, Config,
                                           query(QueryEmb), projected(ProjEmb)),
            E,
            (format('  Invocation error: ~w~n', [E]), fail)
        )
    ->  format('  Projected embedding: ~w~n', [ProjEmb]),
        assert_true(is_list(ProjEmb), 'projection returns list'),
        length(ProjEmb, Len),
        assert_true(Len =:= 4, 'projection has correct dimension')
    ;   format('  X Projection invocation FAILED~n'),
        fail
    ),

    % Test similarity computation
    DocEmb = [1.0, 0.0, 0.0, 0.0],
    (   catch(
            lda_projection:invoke_component(test_projection, Config,
                                           similarity(QueryEmb, DocEmb), score(Score)),
            E2,
            (format('  Similarity error: ~w~n', [E2]), fail)
        )
    ->  format('  Similarity score: ~w~n', [Score]),
        assert_true(number(Score), 'similarity returns number'),
        assert_true(Score > 0.99, 'identical vectors have high similarity')
    ;   format('  X Similarity invocation FAILED~n'),
        fail
    ),

    % Cleanup: remove test component
    retract_component(runtime, test_projection),

    format('~n').

%% ============================================
%% Main
%% ============================================

:- initialization(run_tests, main).
